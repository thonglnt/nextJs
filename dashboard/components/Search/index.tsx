import {
  Dialog,
  DialogTitle,
  Paper,
  Typography,
  IconButton,
  Box,
  useTheme,
} from '@mui/material'
import { memo } from 'react'
import CloseIcon from '@mui/icons-material/Close'

// Components
import Input from '@components/Input'

export interface ISearchDialog {
  isOpen: boolean
  onClose?: () => void
}

const SearchDialog = ({ isOpen = false, onClose }: ISearchDialog) => {
  const theme = useTheme()

  return (
    <Dialog
      open={isOpen}
      onClose={onClose}
      aria-labelledby="title"
      data-testid="modalSearch"
    >
      <Paper
        sx={{
          width: '600px',
          maxWidth: '600px',
        }}
      >
        <DialogTitle
          id="title"
          display="flex"
          justifyContent="space-between"
          bgcolor={theme.palette.secondary.main}
        >
          <Typography
            color={theme.palette.common.white}
            fontSize={16}
            lineHeight="24px"
          >
            Search
          </Typography>
          <IconButton
            size="small"
            edge="start"
            color="inherit"
            onClick={onClose}
          >
            <CloseIcon
              sx={{
                width: '20px',
                height: '20px',
                color: theme.palette.common.white,
              }}
            />
          </IconButton>
        </DialogTitle>
        <Box
          sx={{
            padding: '20px 24px',
          }}
        >
          <Input name="input" label="Search" fullWidth />
        </Box>
      </Paper>
    </Dialog>
  )
}

export default memo(SearchDialog)
