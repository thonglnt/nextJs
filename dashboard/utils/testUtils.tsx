import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'

// Themes
import theme from '@self-types/Themes.types'

export const customRender = (children: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
}

export * from '@testing-library/react'
export { customRender as render }
