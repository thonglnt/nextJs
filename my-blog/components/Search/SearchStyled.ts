import { Modal, styled } from '@mui/material'

export const BackDropStyled = styled(Modal)`
  position: fixed;
  inset: 0;
  display: flex;
  align-items: center;
  justify-content: center;
`
