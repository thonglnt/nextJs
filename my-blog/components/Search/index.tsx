'use client'

import { Box, Input, SvgIcon, Typography, useTheme } from '@mui/material'
import { ChangeEvent, memo, useCallback } from 'react'
import CloseIcon from '@mui/icons-material/Close'

// Styles
import { BackDropStyled } from './SearchStyled'

export interface ISearch {
  isOpen: boolean
  onClose: () => void
  onChange: (e: ChangeEvent<HTMLInputElement>) => void
  onKeyDown: (e: React.KeyboardEvent<HTMLInputElement>) => void
}
const Search = ({ isOpen, onClose, onChange, onKeyDown }: ISearch) => {
  const theme = useTheme()

  const preventCloseModal = useCallback(
    (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => e.stopPropagation(),
    [],
  )

  if (!isOpen) {
    return null
  }

  return (
    <BackDropStyled
      open={isOpen}
      onClose={onClose}
      aria-labelledby="modal-modal-title"
      aria-describedby="modal-modal-description"
      data-testid="backdrop"
    >
      <Box
        data-testid="content"
        onClick={preventCloseModal}
        bgcolor={theme.palette.common.black}
        padding="15px"
        position="relative"
        sx={{
          width: {
            xs: '90%',
            md: '50%',
          },
        }}
      >
        <SvgIcon
          data-testid="close-icon"
          sx={{
            color: theme.palette.common.white,
            position: 'absolute',
            right: 0,
            top: 0,
            cursor: 'pointer',
          }}
          onClick={onClose}
        >
          <CloseIcon />
        </SvgIcon>
        <Input
          type="text"
          placeholder="Search ..."
          fullWidth
          sx={{
            color: theme.palette.common.white,
            fontSize: '30px',
            border: 'none',
            marginTop: '10px',
          }}
          onChange={onChange}
          onKeyDown={onKeyDown}
        />
        <Typography
          fontSize="13px"
          sx={{
            color: theme.palette.common.white,
            fontWeight: 600,
            marginTop: '5px',
          }}
        >
          Type to search
        </Typography>
      </Box>
    </BackDropStyled>
  )
}

export default memo(Search)
