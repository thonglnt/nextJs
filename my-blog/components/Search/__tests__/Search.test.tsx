// Utils
import { render, fireEvent, screen } from '@utils/testUtils'
import Search from '..'

// Components

describe('Search component', () => {
  const props = {
    isOpen: true,
    onClose: jest.fn(),
    onChange: jest.fn(),
    onKeyDown: jest.fn(),
  }

  it('should render Search', () => {
    const { container } = render(<Search {...props} />)

    expect(container).toMatchSnapshot()
  })

  it('should call onChange Search ', () => {
    render(<Search {...props} />)

    const search = screen.getByRole('textbox') as HTMLInputElement

    fireEvent.change(search, {
      target: { value: 'Search Value' },
    })

    expect(props.onChange).toHaveBeenCalled()
    expect(search.value).toBe('Search Value')
  })

  it('should call onKeyDown after Enter ', () => {
    render(<Search {...props} />)

    const search = screen.getByRole('textbox')
    fireEvent.keyDown(search)

    expect(props.onKeyDown).toHaveBeenCalled()
  })

  it('should call onClose after click icon close ', () => {
    render(<Search {...props} />)

    const closeIcon = screen.getByTestId('close-icon')
    fireEvent.click(closeIcon)

    expect(props.onClose).toHaveBeenCalled()
  })

  it('should call onClose after click icon close ', () => {
    render(<Search {...props} />)

    const backdrop = screen.getByTestId('backdrop')
    fireEvent.click(backdrop)

    expect(props.onClose).toHaveBeenCalled()
  })

  it('do not render modal when it is not open', async () => {
    render(<Search {...props} isOpen={false} />)

    const content = screen.queryByTestId('content')
    expect(content).not.toBeTruthy()
  })
})
