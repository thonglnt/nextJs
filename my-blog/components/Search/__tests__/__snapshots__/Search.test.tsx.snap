// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should render Search 1`] = `
<div
  aria-hidden="true"
/>
`;
